Repo URL: https://github.com/<redacted>/assignment-5-judging-flowers-<redacted>

Commit ID: c0a223d

In [131]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [132]:
# Load the dataset (load remotely or locally)
data = pd.read_csv('iris.csv')
# Output the first 15 rows of the data
print("First 15 rows")
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
print('\nSummary')
print(data.describe())

First 15 rows
    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data is 4 measurements taken from flowers and their species. The measurements/features are sepal width, sepal length, petal length, and petal width. Their species/labels are setosa, versicolor, and verginica. Setosa corresponds with Iris-Setosa, versicolor corresponds with Iris-Versicolor, and verginica corresponds with Iris-Virginica.

# Part 2: Split the dataset into train and test

In [133]:
# Take the dataset and split it into our features (X) and label (y)
X = data.drop(columns=['species'])
y = data['species']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Part 3: Logistic Regression

In [134]:
# i. Use sklearn to train a LogisticRegression model on the training set
model = LogisticRegression()
model.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = X_test.iloc[0:1]
print('Sample:', sample_datapoint)
predicted_output = model.predict(sample_datapoint)
print('Prediction:', predicted_output)
# iii. Report on the score for Logistic regression model, what does the score measure?
score = model.score(X_test, y_test)
print('Score:', score)
# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = model.coef_
intercept = model.intercept_
print('Intercept:', intercept, '\nCoefficients:', coefficients)

Sample:     sepal_length  sepal_width  petal_length  petal_width
73           6.1          2.8           4.7          1.2
Prediction: ['versicolor']
Score: 1.0
Intercept: [  9.54356486   1.89887279 -11.44243765] 
Coefficients: [[-0.42830334  0.96729205 -2.44742526 -1.03780396]
 [ 0.5130202  -0.22029888 -0.21604909 -0.84453741]
 [-0.08471686 -0.74699317  2.66347435  1.88234137]]


A score of 1 means that the model made the correct prediction for 100% of the data.

# Part 4: Support Vector Machine

In [135]:
# i. Use sklearn to train a Support Vector Classifier on the training set
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)
svm_classifier = SVC(kernel='linear', probability=True)

svm_classifier.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
sample_datapoint = X_test[0:1]
print('Sample:', sample_datapoint)
# sample_datapoint = scaler.transform(sample_datapoint)
probabilities = svm_classifier.predict_proba(sample_datapoint)

# Show the predicted probabilities
print("Predicted probabilities for each class:", probabilities)
# iii. Report on the score for the SVM, what does the score measure?
score = svm_classifier.score(X_test, y_test)
print('Score:', score)

Sample: [[ 0.32111514 -0.58672696  0.51776964 -0.01658915]]
Predicted probabilities for each class: [[0.00678911 0.96426506 0.02894583]]
Score: 0.9333333333333333


The score of 0.933 implies that the SVM made correct predictions for 93.3% of the test data

# Part 5: Neural Network

In [136]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)
mlp.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities = mlp.predict_proba(sample_datapoint)
print("Predicted probabilities for each class:", predicted_probabilities)
# iii. Report on the score for the Neural Network, what does the score measure?
score = mlp.score(X_test, y_test)
print("Model Score (Accuracy):", score)
# iv: Experiment with different options for the neural network, report on your best configuration
# A list of different configurations to test
configs = [
    {'hidden_layer_sizes': (1,), 'activation': 'relu'},
    {'hidden_layer_sizes': (3,), 'activation': 'relu'},
    {'hidden_layer_sizes': (3,3), 'activation': 'relu'},
    {'hidden_layer_sizes': (10,), 'activation': 'relu'},
    {'hidden_layer_sizes': (10,10,10), 'activation': 'relu'},
    {'hidden_layer_sizes': (50,5), 'activation': 'relu'},
    {'hidden_layer_sizes': (100,), 'activation': 'relu'},
    {'hidden_layer_sizes': (50, 50), 'activation': 'relu'},
    {'hidden_layer_sizes': (100,), 'activation': 'tanh'},
    {'hidden_layer_sizes': (100, 50), 'activation': 'logistic'},
    {'hidden_layer_sizes': (30, 30, 30), 'activation': 'logistic'},
]

best_score = 0
best_configuration = None

for config in configs:
    mlp = MLPClassifier(hidden_layer_sizes=config['hidden_layer_sizes'], 
                        activation=config['activation'], 
                        max_iter=1000, 
                        random_state=42)
    mlp.fit(X_train, y_train)
    score = mlp.score(X_test, y_test)
    print(f"Configuration: {config}, Score: {score}")

    if score > best_score:
        best_score = score
        best_configuration = config

print("Best Configuration:", best_configuration)
print("Best Score:", best_score)


Predicted probabilities for each class: [[0.00142997 0.98365823 0.0149118 ]]
Model Score (Accuracy): 1.0


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


Configuration: {'hidden_layer_sizes': (1,), 'activation': 'relu'}, Score: 0.7333333333333333


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


Configuration: {'hidden_layer_sizes': (3,), 'activation': 'relu'}, Score: 0.9333333333333333


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (1000) reached and the optimization hasn't converged yet.
  warnings.warn(


Configuration: {'hidden_layer_sizes': (3, 3), 'activation': 'relu'}, Score: 0.7333333333333333
Configuration: {'hidden_layer_sizes': (10,), 'activation': 'relu'}, Score: 1.0
Configuration: {'hidden_layer_sizes': (10, 10, 10), 'activation': 'relu'}, Score: 0.9333333333333333
Configuration: {'hidden_layer_sizes': (50, 5), 'activation': 'relu'}, Score: 1.0
Configuration: {'hidden_layer_sizes': (100,), 'activation': 'relu'}, Score: 1.0
Configuration: {'hidden_layer_sizes': (50, 50), 'activation': 'relu'}, Score: 1.0
Configuration: {'hidden_layer_sizes': (100,), 'activation': 'tanh'}, Score: 0.9333333333333333
Configuration: {'hidden_layer_sizes': (100, 50), 'activation': 'logistic'}, Score: 0.9333333333333333
Configuration: {'hidden_layer_sizes': (30, 30, 30), 'activation': 'logistic'}, Score: 1.0
Best Configuration: {'hidden_layer_sizes': (10,), 'activation': 'relu'}
Best Score: 1.0


The neural network had a score of 1.0, meaning it predicted correctly for 100% of the test data. A few configurations had a score of 1.0, the simplest one had 10 nodes in its hidden layer and a Rectified Linear Unit activation function.

# Part 6: K-Nearest Neighbors

In [139]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=5)  # You can adjust the number of neighbors (k)
knn.fit(X_train, y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
predicted_probabilities = knn.predict_proba(sample_datapoint)

print("Predicted probabilities for each class:", predicted_probabilities)
# iii. Report on the score for kNN, what does the score measure?
score = knn.score(X_test, y_test)
print("Model Score (Accuracy):", score)

Predicted probabilities for each class: [[0. 1. 0.]]
Model Score (Accuracy): 1.0


The model's score was 1.0, meaning it predicted the correct class 100% of the time

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

All the explored types of models were able to accurately predict the proper class of flower based on the dataset. All the models performed equally well (with good choices for neural net config) except for SVM. It seems this is mostly due to a different seed being chosen for SVM causing it to perform slightly worse since this is a small dataset. I think the most surprising thing about this exercise was how easy it is to carry out complex things like neural networks. The math behind these ideas can be so complex, but we can take advantage of it all without even needing to understand it with only a few lines of code.